File: lib/lua/virgo-time.lua
--[[
Copyright 2012 Rackspace

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS-IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
--]]

local logging = require('logging')
local Error = require('core').Error
local math = require('math')

local delta = 0
local delay = 0

local function now()
  return math.floor(virgo.gmtnow() + delta)
end

local function raw()
  return math.floor(virgo.gmtnow())
end

local function setDelta(_delta)
  delta = _delta
end

local function getDelta()
  return delta
end

--[[

This algorithm follows the NTP algorithm found here:

http://www.eecis.udel.edu/~mills/ntp/html/warp.html

T1 = agent departure timestamp
T2 = server receieved timestamp
T3 = server transmit timestamp
T4 = agent destination timestamp

]]--
local function timesync(T1, T2, T3, T4)
  if not T1 or not T2 or not T3 or not T4 then
    return Error:new('T1, T2, T3, or T4 was null. Failed to sync time.')
  end

  logging.debugf('time_sync data: T1 = %.0f T2 = %.0f T3 = %.0f T4 = %.0f', T1, T2, T3, T4)

  delta = ((T2 - T1) + (T3 - T4)) / 2
  delay = ((T4 - T1) + (T3 - T2))

  logging.infof('Setting time delta to %.0fms based on server time %.0fms', delta, T2)

  return
end

local exports = {}
exports.setDelta = setDelta
exports.getDelta = getDelta
exports.now = now
exports.raw = raw
exports.timesync = timesync
return exports

File: lib/virgo_time.c
/*
 *  Copyright 2012 Rackspace
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
#ifdef _WIN32
#include <time.h>
#else
#include <sys/time.h>
#endif

#include <stdio.h>
#include <math.h>
#include "lua.h"
#include "virgo.h"
#include "virgo_portable.h"

#ifdef _WIN32

#include <windows.h>

union timestamp_u {
  FILETIME ft_;
  int64_t t_;
};

#define kTimeEpoc 116444736000000000LL
#define kTimeScaler 10000

static double
virgo__time_now() {
  union timestamp_u ts;
  GetSystemTimeAsFileTime(&ts.ft_);
  return (double)((ts.t_ - kTimeEpoc) / kTimeScaler);
}

#else

static double
virgo__time_now() {
  struct timeval tv;
  if (gettimeofday(&tv, NULL) < 0) {
    return 0.0;
  }
  return (((double)tv.tv_sec * 1000) + ((double)tv.tv_usec / 1000));
}

#endif

int
virgo_time_now(lua_State *L) {
  lua_pushnumber(L, virgo__time_now());
  return 1;
}
